ADD: tool to generate last mnemonic word

3bitcoins opened the pull request:
Closes #5680 

![Screenshot_1697197785](https://github.com/BlueWallet/BlueWallet/assets/155891/2148fcd8-1d04-4f1d-aece-c9e455d5f615)

![Screenshot_1697197935](https://github.com/BlueWallet/BlueWallet/assets/155891/62369d3f-2341-48c6-a568-50558c87ecdb)

limpbrains commented:
Can you please add tests to the PR. Maybe extract checkMnemonic logic to a separate function and write a unit test for it.

Thank you for contributing to BlueWallet!

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

3bitcoins commented:
@limpbrains I've moved the core logic to it's own file and added tests. Do you guys plan to ever review #5690 ? If so, I'll adapt it to use this helper instead of having this logic implemented twice.

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

limpbrains commented:
I've tried, works fine. 
I think UX can be better. 
- We need to show error if mnemonic is incorrect.
- show last word before button.
- Maybe text explanation of what is going on here whould be nice, user can be confused about why it is random. (I'm not sure)

DrockDaniel commented:
I don't know how to do any of this it was a long time ago I created bitcoin

On Wed, Oct 11, 2023, 4:54 AM Ivan ***@***.***> wrote:

> Can you please add tests to the PR. Maybe extract checkMnemonic logic to a
> separate function and write a unit test for it.
>
> Thank you for contributing to BlueWallet!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BlueWallet/BlueWallet/pull/5722#issuecomment-1757297113>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUOFMFXI2LH2HVP3RRIOXWLX6ZUEDANCNFSM6AAAAAA5VAXBOY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

3bitcoins commented:
@limpbrains I've swapped the place of the result and the button, and added a message if the input is invalid. I also fixed a couple linter warnings i missed.

Based on how the other tools are, I don't think it's necessary to describe the exact usage of the tool in the screen itself. If you want, maybe some kind of documentation?

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

3bitcoins commented:
@Overtorment Do you think you could review this?

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

GladosBlueWallet commented:
Wake the fuck up samurai, we have PRs to merge

![image](https://user-images.githubusercontent.com/1913337/127044132-147fa451-6dc1-454f-907a-9292033047f2.png)

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

Overtorment commented on tests/unit/checksumWords.test.js:
```diff
@@ -0,0 +1,39 @@
+      'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon';
+    const result = generateChecksumWords(input);
+    assert.ok(result);
+    assert.strictEqual(result.length, 8);
```
would be nice to check actual words generated

3bitcoins replied:
logically it seems that if the right amount are generated and they're all valid, it's generated the right words

Overtorment commented on screen/wallets/generateWord.js:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    const random = await randomBytes(1);
+    const randomindex = Math.round((random.readUInt8(0) / 255) * (possibleWords.length - 1));
```
`Math.random()` might have worked better here, without this awkward buffer-to-int conversion (`readUInt8(0) / 255`). not a deal breaker tho

3bitcoins replied:
I wasn't sure to what extent Math.random() was considered cryptographically secure in this project, and opted to use the safer RNG that i see used for other keygen purposes

Overtorment commented on blue_modules/checksumWords.js:
```diff
@@ -0,0 +1,79 @@
+import * as bip39 from 'bip39';
+import createHash from 'create-hash';
+
+// partial (11 or 23 word) seed phrase
```
link to documentation would be nice

Overtorment reviewed (approved):
thanks for the contribution! merging this.
however,  theres room for improvement,  if you fancy doing a follow-up PR, most importantly, converting it to typescript (we now have a rule - every new file added might be in ts, while old js files are gradually converted)
